### PR TITLE
Hasselblad X2D 100C support

### PR DESCRIPTION
**Pull request by kmilos**
Addresses https://github.com/darktable-org/darktable/issues/14345

Stub only (using ADC 15.3), levels and crop need to be verified.

Depends on #466 for OOC 3FR uncompressed support.

### CODE REVIEW

**Comment by kmilos** [comment]
Note also that the 3FR has these tags (does unique ID matter?):
```
UniqueCameraModel               : Hasselblad X2D 100C
ColorMatrix1                    : 0.5950573751 -0.146674752 -0.03412952103 -0.5159296309 1.26306183 0.2826110949 -0.103825823 0.1675577691 0.6152086
```
but no idea what that matrix is calibrated for... I presume older Hasselblad models also use the Adobe one instead of the 3FR embedded one? (They also seem to use the Adobe unique ID instead of the 3FR one... And this doesn't seem to matter as canonical_id is not used by dt currently...)

**Comment by kkempter-getswift** [comment]
I pulled an image from the X2d into lightroom and then exported it t a DNG file. 

Here is the BlackLevel info from the resulting DNG file:

$ exiv2 -pt B0000050.dng| grep BlackLevel
Exif.SubImage1.BlackLevel                    Rational    1  4267/1
Exif.SubImage3.BlackLevel                    Rational    3  51/1 51/1 51/1
Exif.SubImage4.BlackLevel                    Rational    3  51/1 51/1 51/1
Exif.SubImage5.BlackLevel                    Rational    3  51/1 51/1 51/1


I'm not sure what to do with this, how/where to update the xml file so I just added it here in case it helps

**Comment by kkempter-getswift** [comment]
and here are the color matrix info:

$ exiv2 -pt B0000050.dng| grep CalibrationIlluminant2
Exif.Image.CalibrationIlluminant2            Short       1  D65

$ exiv2 -pt B0000050.dng| grep ColorMatrix2
Exif.Image.ColorMatrix2                      SRational   9  6468/10000 -1899/10000 -545/10000 -4526/10000 12267/10000 2542/10000 -388/10000 1276/10000 6096/10000


from the same DNG file (see my comment above)

**Comment by LebedevRI** [comment]
> $ exiv2 -pt B0000050.dng| grep CalibrationIlluminant2 Exif.Image.CalibrationIlluminant2 Short 1 D65
> 
> $ exiv2 -pt B0000050.dng| grep ColorMatrix2 Exif.Image.ColorMatrix2 SRational 9 6468/10000 -1899/10000 -545/10000 -4526/10000 12267/10000 2542/10000 -388/10000 1276/10000 6096/10000

So we can infer/guess that at least for that camera the camera-provided color matrix is D65, i guess.

**Comment by kmilos** [comment]
@LebedevRI Looks like we'll also need to bump up the accepted size a bit: https://discuss.pixls.us/t/looking-for-workflow-advice-with-3fr-files-from-a-hasselblad-x2d/36897/27

H6D we already had on RPU is 12000 x 8816, new X2D is 11904 x 8842

**Comment by kofa73** [comment]
With darktable master + your PR:

    kofa@eagle:~/darktable/src/external/rawspeed$ git status
    On branch hassy_x2d
    nothing to commit, working tree clean
    kofa@eagle:~/darktable/src/external/rawspeed$ git rev-parse HEAD
    7fd3ca0e0d7d6fa1e2a504d9b94a014861a7fd9f

    kofa@eagle:~/darktable$ git status
    On branch master
    Your branch is up to date with 'origin/master'.
    
    Changes not staged for commit:
      (use "git add <file>..." to update what will be committed)
      (use "git restore <file>..." to discard changes in working directory)
            modified:   src/external/rawspeed (new commits)
    
    no changes added to commit (use "git add" and/or "git commit -a")
    kofa@eagle:~/darktable$ git rev-parse HEAD
    aaf481c2aee61b4c2b9eca90ea95a15787e31d3e


Opening the raw from https://nextcloud.phxicg.com/s/9eGHMdPc8Q6jTEa, I get:

    92.6376 [rawspeed] (B0000050.3FR) rawspeed::JpegMarker rawspeed::AbstractLJpegDecoder::getNextMarker(bool), line 299: (Noskip) Expected marker not found. Probably corrupt file.
    92.6376 [libraw_open] extensions whitelist: `cr3'
    92.6377 [dt_dev_load_raw] loading the image. took 0.404 secs (0.000 CPU)
    92.9398 [temperature] failed to read camera white balance information from `B0000050.3FR'!
    92.9482 Trouble: [lens correction] camera/lens not found (B0000050.3FR 24873)
    92.9542 [colorin] could not find requested profile `standard color matrix'!
    93.0787 [histogram] took 0.000 secs (0.000 CPU) scope draw

**Comment by kmilos** [comment]
That is pretty weird as the RPU sample is uncompressed.

@kkempter-getswift Does the camera support both uncompressed and compressed raw output? Please upload all possible outputs to RPU, as requested there.

**Comment by LebedevRI** [comment]
> 92.6376 [rawspeed] (B0000050.3FR) rawspeed::JpegMarker rawspeed::AbstractLJpegDecoder::getNextMarker(bool), line 299: (Noskip) Expected marker not found. Probably corrupt file.

I haven't checked, but that sounds like #144.

**Comment by kkempter-getswift** [comment]
Hi everyone

I could use some help. I am a DBA and less of a developer, I use git & friends but minimally and even then I am quite rusty.

I was the one who uploaded the .3FR files

I downloaded 4.2.1 source code from https://www.darktable.org/install/ and compiled on my local Fedora laptop.

Now I want to grab the source code that would include this PR and re-test.  Can someone help me with how to grab the right code?

Thanks in advance

**Comment by kmilos** [comment]
> I haven't checked, but that sounds like #144.

But it should be uncompressed! And the size seems to corroborate that - 2*100MB...

**Comment by LebedevRI** [comment]
> > I haven't checked, but that sounds like #144.
> 
> But it should be uncompressed! And the size seems to corroborate that - 2*100MB...

I'm not sure what you mean?
For 3FR, we don't even have non-LJpeg options: https://github.com/darktable-org/rawspeed/blob/a02d62ecfa1fe482884b6168abf4d540d7edb70a/src/librawspeed/decoders/ThreefrDecoder.cpp#L52-L69

**Comment by kkempter-getswift** [comment]
per this request:

That is pretty weird as the RPU sample is uncompressed.

@kkempter-getswift Does the camera support both uncompressed and compressed raw output? Please upload all possible outputs to RPU, as requested there.

The camera allows me to select either RAW or "RAW + JPG"

and then I can choose 14bit or 16bit

there are no other options. The files I uploaded are 16bit

I just uploaded a 14bit example file (B0000077.3FR)

**Comment by kmilos** [comment]
> I'm not sure what you mean?
> For 3FR, we don't even have non-LJpeg options

Ok, that would explain it then 😉

At least this 16b X2D sample is uncompressed so needs to be hooked up differently.

**Comment by kmilos** [comment]
> I just uploaded a 14bit example file (B0000077.3FR)

There is no difference whatsoever between the 14-bit and 16-bit modes it would seem - still uncompressed, and data is 16-bit, with no gaps in the values.

**Comment by kmilos** [comment]
Also, the X1D II 50C samples on RPU are uncompressed, and I think [supported](https://github.com/darktable-org/rawspeed/commit/bf07cc438e1c8344bd141fdafc7109a34e34d531)? @LebedevRI 

Edit: Looks like there is also a compressed X1DM2-50c (.fff) sample, and the X1D II 50C (.3FR) uncompressed one. Only compressed mode was checked perhaps? So [this report](https://github.com/darktable-org/darktable/issues/10465) then makes more sense?

Yep, out of e.g. these
image
only the compressed X1DM2-50c works, while the top three seem to be the identical "X1D II 50C" body. (No idea where the "cfv 50C II" duplicate came from, it seems to have nothing like that in the metadata...)

**Comment by kmilos** [comment]
Also added (copied w/ crop tweak for Mk II as well) support for the X1D while we're at it since the sample is there.

**Comment by LebedevRI** [comment]
Please split this into 3+ sequential PR's - uncompressed, and then per camera.

**Comment by kmilos** [comment]
> Please split this into 3+ sequential PR's - uncompressed, and then per camera.

Done.

This now depends on https://github.com/darktable-org/rawspeed/pull/466

File: data/cameras.xml [hunk 1]
```diff
@@ -17159,6 +17159,24 @@
...
 	</Camera>
+	<Camera make="Hasselblad" model="X2D 100C">
+		<ID make="Hasselblad" model="X2D 100C">Hasselblad 100-20-Coated6</ID>
+		<CFA width="2" height="2">
+			<Color x="0" y="0">RED</Color>
+			<Color x="1" y="0">GREEN</Color>
+			<Color x="0" y="1">GREEN</Color>
+			<Color x="1" y="1">BLUE</Color>
+		</CFA>
+		<Crop x="126" y="94" width="-118" height="0"/>
```
Comment (LebedevRI) [inline data/cameras.xml:17170]: Same, crop still leaves a few garbage pixels in.
How do you come up with it? Have you tried the methodology i outlined in some previous PR?

**Comment by LebedevRI** [comment]
Thank you!

**Comment by kkempter-getswift** [comment]
I suspect you all have fixed this already, and I suspect I'm doing something wrong to get the latest code with the x2d fixes, however just in case it's helpful:

I checked out the master branch and compiled

I also tried via checking out the master branch then:
~/darktable$ cd src/external/rawspeed/
~/darktable/src/external/rawspeed$ gh pr checkout 463
and compiled


In both cased above I get this when I open darktable, import my 3FR images and then double click on one of the images:


     **2.5480 [rawspeed] (B0000060.3FR) rawspeed::HasselbladLJpegDecoder::HasselbladLJpegDecoder(rawspeed::ByteStream, const rawspeed::RawImage&), line 43: Unexpected image dimensions found: (11904; 8842)
     2.5480 [libraw_open] extensions whitelist: `cr3'
     2.8348 [temperature] failed to read camera white balance information from `B0000060.3FR'!**

File: data/cameras.xml (hunk 1 above)
Comment (kmilos) [reply data/cameras.xml:17170]: Sorry, I indeed had no time to go through the zoom & scroll procedure, this was from raw data PPM dump...

**Comment by kmilos** [comment]
@kkempter-getswift All the necessary changes have been merged now, so you can just pull the develop branch again, no need for individual PR checkout...

Also, as mentioned, if possible, please upload the equivalent .FFF file to RPU please.

**Comment by kkempter-getswift** [comment]
Thanks,

I dont see any .FFF files, when I connect the camera as a storage device I have a DCIM directory and in that directory a sub directory (this one is named 114HASBL) and within the sub directory I have only the .3FR files, here is a directory listing:

file:///run/media/kevink/X2D-100C-CF/DCIM/114HASBL/B0000004.3FR
file:///run/media/kevink/X2D-100C-CF/DCIM/114HASBL/B0000005.3FR
file:///run/media/kevink/X2D-100C-CF/DCIM/114HASBL/B0000006.3FR
file:///run/media/kevink/X2D-100C-CF/DCIM/114HASBL/B0000007.3FR
file:///run/media/kevink/X2D-100C-CF/DCIM/114HASBL/B0000008.3FR
file:///run/media/kevink/X2D-100C-CF/DCIM/114HASBL/B0000009.3FR
file:///run/media/kevink/X2D-100C-CF/DCIM/114HASBL/B0000010.3FR
file:///run/media/kevink/X2D-100C-CF/DCIM/114HASBL/B0000011.3FR
file:///run/media/kevink/X2D-100C-CF/DCIM/114HASBL/B0000012.3FR
file:///run/media/kevink/X2D-100C-CF/DCIM/114HASBL/B0000013.3FR
file:///run/media/kevink/X2D-100C-CF/DCIM/114HASBL/B0000014.3FR
file:///run/media/kevink/X2D-100C-CF/DCIM/114HASBL/B0000015.3FR
file:///run/media/kevink/X2D-100C-CF/DCIM/114HASBL/B0000016.3FR
file:///run/media/kevink/X2D-100C-CF/DCIM/114HASBL/B0000017.3FR
file:///run/media/kevink/X2D-100C-CF/DCIM/114HASBL/B0000018.3FR
file:///run/media/kevink/X2D-100C-CF/DCIM/114HASBL/B0000019.3FR
file:///run/media/kevink/X2D-100C-CF/DCIM/114HASBL/B0000020.3FR
file:///run/media/kevink/X2D-100C-CF/DCIM/114HASBL/B0000021.3FR
file:///run/media/kevink/X2D-100C-CF/DCIM/114HASBL/B0000022.3FR
file:///run/media/kevink/X2D-100C-CF/DCIM/114HASBL/B0000023.3FR
file:///run/media/kevink/X2D-100C-CF/DCIM/114HASBL/B0000024.3FR
file:///run/media/kevink/X2D-100C-CF/DCIM/114HASBL/B0000025.3FR
file:///run/media/kevink/X2D-100C-CF/DCIM/114HASBL/B0000026.3FR
file:///run/media/kevink/X2D-100C-CF/DCIM/114HASBL/B0000027.3FR
file:///run/media/kevink/X2D-100C-CF/DCIM/114HASBL/B0000028.3FR
file:///run/media/kevink/X2D-100C-CF/DCIM/114HASBL/B0000029.3FR
file:///run/media/kevink/X2D-100C-CF/DCIM/114HASBL/B0000030.3FR
file:///run/media/kevink/X2D-100C-CF/DCIM/114HASBL/B0000031.3FR
file:///run/media/kevink/X2D-100C-CF/DCIM/114HASBL/B0000032.3FR
file:///run/media/kevink/X2D-100C-CF/DCIM/114HASBL/B0000033.3FR
file:///run/media/kevink/X2D-100C-CF/DCIM/114HASBL/B0000034.3FR
file:///run/media/kevink/X2D-100C-CF/DCIM/114HASBL/B0000035.3FR
file:///run/media/kevink/X2D-100C-CF/DCIM/114HASBL/B0000036.3FR
file:///run/media/kevink/X2D-100C-CF/DCIM/114HASBL/B0000037.3FR
file:///run/media/kevink/X2D-100C-CF/DCIM/114HASBL/B0000038.3FR
file:///run/media/kevink/X2D-100C-CF/DCIM/114HASBL/B0000039.3FR
file:///run/media/kevink/X2D-100C-CF/DCIM/114HASBL/B0000040.3FR
file:///run/media/kevink/X2D-100C-CF/DCIM/114HASBL/B0000041.3FR
file:///run/media/kevink/X2D-100C-CF/DCIM/114HASBL/B0000042.3FR
file:///run/media/kevink/X2D-100C-CF/DCIM/114HASBL/B0000043.3FR
file:///run/media/kevink/X2D-100C-CF/DCIM/114HASBL/B0000044.3FR
file:///run/media/kevink/X2D-100C-CF/DCIM/114HASBL/B0000045.3FR
file:///run/media/kevink/X2D-100C-CF/DCIM/114HASBL/B0000046.3FR
file:///run/media/kevink/X2D-100C-CF/DCIM/114HASBL/B0000047.3FR
file:///run/media/kevink/X2D-100C-CF/DCIM/114HASBL/B0000048.3FR
file:///run/media/kevink/X2D-100C-CF/DCIM/114HASBL/B0000049.3FR
file:///run/media/kevink/X2D-100C-CF/DCIM/114HASBL/B0000050.3FR
file:///run/media/kevink/X2D-100C-CF/DCIM/114HASBL/B0000051.3FR
file:///run/media/kevink/X2D-100C-CF/DCIM/114HASBL/B0000052.3FR
file:///run/media/kevink/X2D-100C-CF/DCIM/114HASBL/B0000053.3FR
file:///run/media/kevink/X2D-100C-CF/DCIM/114HASBL/B0000054.3FR
file:///run/media/kevink/X2D-100C-CF/DCIM/114HASBL/B0000055.3FR
file:///run/media/kevink/X2D-100C-CF/DCIM/114HASBL/B0000056.3FR
file:///run/media/kevink/X2D-100C-CF/DCIM/114HASBL/B0000057.3FR
file:///run/media/kevink/X2D-100C-CF/DCIM/114HASBL/B0000058.3FR
file:///run/media/kevink/X2D-100C-CF/DCIM/114HASBL/B0000059.3FR
file:///run/media/kevink/X2D-100C-CF/DCIM/114HASBL/B0000060.3FR
file:///run/media/kevink/X2D-100C-CF/DCIM/114HASBL/B0000061.3FR
file:///run/media/kevink/X2D-100C-CF/DCIM/114HASBL/B0000062.3FR
file:///run/media/kevink/X2D-100C-CF/DCIM/114HASBL/B0000063.3FR
file:///run/media/kevink/X2D-100C-CF/DCIM/114HASBL/B0000064.3FR
file:///run/media/kevink/X2D-100C-CF/DCIM/114HASBL/B0000065.3FR
file:///run/media/kevink/X2D-100C-CF/DCIM/114HASBL/B0000066.3FR
file:///run/media/kevink/X2D-100C-CF/DCIM/114HASBL/B0000067.3FR
file:///run/media/kevink/X2D-100C-CF/DCIM/114HASBL/B0000068.3FR
file:///run/media/kevink/X2D-100C-CF/DCIM/114HASBL/B0000069.3FR
file:///run/media/kevink/X2D-100C-CF/DCIM/114HASBL/B0000070.3FR
file:///run/media/kevink/X2D-100C-CF/DCIM/114HASBL/B0000071.3FR
file:///run/media/kevink/X2D-100C-CF/DCIM/114HASBL/B0000072.3FR
file:///run/media/kevink/X2D-100C-CF/DCIM/114HASBL/B0000073.3FR
file:///run/media/kevink/X2D-100C-CF/DCIM/114HASBL/B0000074.3FR
file:///run/media/kevink/X2D-100C-CF/DCIM/114HASBL/B0000075.3FR
file:///run/media/kevink/X2D-100C-CF/DCIM/114HASBL/B0000076.3FR
file:///run/media/kevink/X2D-100C-CF/DCIM/114HASBL/B0000077.3FR
file:///run/media/kevink/X2D-100C-CF/DCIM/114HASBL/B0000078.3FR


If there is something else I can provide let me know, or if I am doing it wrong

Thanks in advance

**Comment by kmilos** [comment]
> If there is something else I can provide let me know, or if I am doing it wrong

I think you get .FFF only if you transfer from camera using Phocus.

Please test transferring both a 16-bit and a 14-bit file - i.e. for the exact two 3FR files you uploaded to RPU (B0000001 and B0000077), create the matching FFF ones by transferring them in Phocus please.

**Comment by kkempter-getswift** [comment]
I just uploaded to RPU a file named Job_0002.fff it was created by the Hasselblad Phocus software when I imported B0000050.3FR I think I already uploaded B0000050.3FR but I will do so again to be sure

**Comment by LebedevRI** [comment]
No `B0000050.3FR` on RPU.

https://github.com/darktable-org/rawspeed/pull/469#issuecomment-1532877586
> @kkempter-getswift could you please follow the guideline outlined on the RPU, i.e. please provide all the nessesary samples (i.e. all 4?, including re-shooing existing samples) from said camera of the same scene, ideally something like daylight landscape?

**Comment by kkempter-getswift** [comment]
I have uploaded the below set of files, each of these are more in line with the guidelines:

B0000009.3FR   and   Job_0009.fff

**Comment by kkempter-getswift** [comment]
I have also uploaded the below set of files, each of these are more in line with the guidelines:

B0000010.3FR and Job_0010.fff

**Comment by LebedevRI** [comment]
Just so we are at the ideal conditions, could you please upload **3** samples //with same content//:
* 16bit uncompressed
* 14bit (uncompressed)
* the `*.fff` produced from the 16bit uncompressed sample
?

**Comment by kkempter-getswift** [comment]
I have uploaded the following 4 files:

B0000079.3FR -> Job_0079.fff (16bit)

B0000080.3FR -> Job_0080.fff (14bit)

**Comment by LebedevRI** [comment]
@kkempter-getswift thank you, verified the uploads!

**Comment by kkempter-getswift** [comment]
All

I compiled the latest development branch, everything is working, Thank You!!!

I am however seeing these output errors on the command line:

    80.4232 Trouble: [color calibration] white balance error (B0000020.3FR 17)
    81.4348 Trouble: [color calibration] white balance error (B0000036.3FR 33)
    81.4432 Trouble: [color calibration] white balance error (B0000036.3FR 33)
    81.4571 Trouble: [white balance] double application of white balance (B0000036.3FR 33)
    81.4821 Trouble: [color calibration] white balance error (B0000036.3FR 33)
    84.2032 Trouble: [color calibration] white balance error (B0000036.3FR 33)
    84.4754 Trouble: [color calibration] white balance error (B0000036.3FR 33)
    84.4755 Trouble: [white balance] double application of white balance (B0000036.3FR 33)
    84.4884 Trouble: [white balance] double application of white balance (B0000036.3FR 33)
    84.5225 Trouble: [color calibration] white balance error (B0000036.3FR 33)
    84.7614 Trouble: [color calibration] white balance error (B0000036.3FR 33)

**Comment by LebedevRI** [comment]
Nice!
That question is more suited for appropriate darktable communication channels.

**Pull request merged**